### PR TITLE
Postgres: prepend className to unique indexes

### PR DESCRIPTION
**Pull request by cbaker6**
When using the PostgresStorageAdapter.ensureUniqueness, the current index name is always "unique_fieldName". This isn't an issue as long as an index field name isn't contained in another table, but causes issues if you want to create a unique index for "uuid" for two different tables. I deleted the comments that mentioned postgres allows the same index name multiple times. I don't know if that was true for older versions of postgres (it doesn't work on postgres 10, 11, and 12), but even the autogenerated primaryKey unique index prepends the table name. The update in this PR should work for older and newer.

The [Uniqueness.spec.js](https://github.com/parse-community/parse-server/blob/d83a0b680854f52ce30205d3558c0133d7f8cf19/spec/Uniqueness.spec.js) makes sure the indexes are created. *Update: I added a test case to make sure multiple unique indexes can be added to different classes that have the same field.*

Probably need to add a recommendation: After upgrading to this version of parse-server, you should log into postgres and drop the old constraints. This should be done manually as it can block writes to the database:

```sql
ALTER TABLE "_User" DROP CONSTRAINT IF EXISTS unique_username;
ALTER TABLE "_User" DROP CONSTRAINT IF EXISTS unique_email;
ALTER TABLE "_Role" DROP CONSTRAINT IF EXISTS unique_name;
```

The above way seems to be recommended by [Heroku](https://devcenter.heroku.com/articles/postgresql-indexes), as this is essentially is reindexing `unique_username`, `unique_email`,  and `unique_name` without blocking writes on the table. Below is a blurb from the link:

> However be cautious about reindexing big indexes as write locks are obtained on the parent table. One strategy to achieve the same result on a live site is to build an index concurrently on the same table and columns but with a different name, and then dropping the original index and renaming the new one. This procedure, while much longer, won’t require any long running locks on the live tables.

### CODE REVIEW

**Comment by cbaker6** [comment]
This should pass after rerunning the test.

It's also ready for review.

**Comment by cbaker6** [comment]
Added another small fix to get rid of console warnings due to attempting to create an index every time during a restart of Postgres. The fix simply adds, `IF NOT EXISTS" when attempting to create indexes, just like `CREATE TABLE IF NOT EXISTS`. The error messages that show up during restart are:

```
db_1         | 2020-06-24 01:36:32.197 UTC [40] ERROR:  relation "unique_username" already exists
db_1         | 2020-06-24 01:36:32.197 UTC [40] STATEMENT:  ALTER TABLE "_User" ADD CONSTRAINT "unique_username" UNIQUE ("username")
db_1         | 2020-06-24 01:36:32.201 UTC [38] ERROR:  relation "case_insensitive_email" already exists
db_1         | 2020-06-24 01:36:32.201 UTC [38] STATEMENT:  CREATE INDEX "case_insensitive_email" ON "_User" (lower("email") varchar_pattern_ops)
db_1         | 2020-06-24 01:36:32.203 UTC [40] ERROR:  relation "unique_name" already exists
db_1         | 2020-06-24 01:36:32.203 UTC [40] STATEMENT:  ALTER TABLE "_Role" ADD CONSTRAINT "unique_name" UNIQUE ("name")
db_1         | 2020-06-24 01:36:32.209 UTC [41] ERROR:  relation "case_insensitive_username" already exists
db_1         | 2020-06-24 01:36:32.209 UTC [41] STATEMENT:  CREATE INDEX "case_insensitive_username" ON "_User" (lower("username") varchar_pattern_ops)
db_1         | 2020-06-24 01:36:32.219 UTC [42] ERROR:  relation "unique_email" already exists
db_1         | 2020-06-24 01:36:32.219 UTC [42] STATEMENT:  ALTER TABLE "_User" ADD CONSTRAINT "unique_email" UNIQUE ("email")
```

**Comment by cbaker6** [comment]
Fixed error messages outputting on restarts. There are still a set of errors below that show on initial start, but I believe those are okay and I don't know how to fix them. From what I see online, it seems this is due to the driver (node-postgres) having multiple connections and issuing the same command over multiple threads. It looks like postgres creates what's necessary and then throws the error from the other threads (I'm guessing). Not a big deal IMO

```
db_1         | 2020-06-24 15:36:34.784 UTC [105] ERROR:  duplicate key value violates unique constraint "pg_type_typname_nsp_index"
db_1         | 2020-06-24 15:36:34.784 UTC [105] DETAIL:  Key (typname, typnamespace)=(_SCHEMA, 2200) already exists.
db_1         | 2020-06-24 15:36:34.784 UTC [105] STATEMENT:  CREATE TABLE IF NOT EXISTS "_SCHEMA" ( "className" varChar(120), "schema" jsonb, "isParseClass" bool, PRIMARY KEY ("className") )
db_1         | 2020-06-24 15:36:34.784 UTC [106] ERROR:  duplicate key value violates unique constraint "pg_type_typname_nsp_index"
db_1         | 2020-06-24 15:36:34.784 UTC [106] DETAIL:  Key (typname, typnamespace)=(_SCHEMA, 2200) already exists.
db_1         | 2020-06-24 15:36:34.784 UTC [106] STATEMENT:  CREATE TABLE IF NOT EXISTS "_SCHEMA" ( "className" varChar(120), "schema" jsonb, "isParseClass" bool, PRIMARY KEY ("className") )
db_1         | 2020-06-24 15:36:34.785 UTC [107] ERROR:  duplicate key value violates unique constraint "pg_type_typname_nsp_index"
db_1         | 2020-06-24 15:36:34.785 UTC [107] DETAIL:  Key (typname, typnamespace)=(_SCHEMA, 2200) already exists.
db_1         | 2020-06-24 15:36:34.785 UTC [107] STATEMENT:  CREATE TABLE IF NOT EXISTS "_SCHEMA" ( "className" varChar(120), "schema" jsonb, "isParseClass" bool, PRIMARY KEY ("className") )
db_1         | 2020-06-24 15:36:34.786 UTC [104] ERROR:  duplicate key value violates unique constraint "pg_type_typname_nsp_index"
db_1         | 2020-06-24 15:36:34.786 UTC [104] DETAIL:  Key (typname, typnamespace)=(_SCHEMA, 2200) already exists.
db_1         | 2020-06-24 15:36:34.786 UTC [104] STATEMENT:  CREATE TABLE IF NOT EXISTS "_SCHEMA" ( "className" varChar(120), "schema" jsonb, "isParseClass" bool, PRIMARY KEY ("className") )
db_1         | 2020-06-24 15:36:34.786 UTC [108] ERROR:  duplicate key value violates unique constraint "pg_type_typname_nsp_index"
db_1         | 2020-06-24 15:36:34.786 UTC [108] DETAIL:  Key (typname, typnamespace)=(_SCHEMA, 2200) already exists.
db_1         | 2020-06-24 15:36:34.786 UTC [108] STATEMENT:  CREATE TABLE IF NOT EXISTS "_SCHEMA" ( "className" varChar(120), "schema" jsonb, "isParseClass" bool, PRIMARY KEY ("className") )
db_1         | 2020-06-24 15:36:34.787 UTC [109] ERROR:  duplicate key value violates unique constraint "pg_type_typname_nsp_index"
db_1         | 2020-06-24 15:36:34.787 UTC [109] DETAIL:  Key (typname, typnamespace)=(_SCHEMA, 2200) already exists.
db_1         | 2020-06-24 15:36:34.787 UTC [109] STATEMENT:  CREATE TABLE IF NOT EXISTS "_SCHEMA" ( "className" varChar(120), "schema" jsonb, "isParseClass" bool, PRIMARY KEY ("className") )
db_1         | 2020-06-24 15:36:34.788 UTC [110] ERROR:  duplicate key value violates unique constraint "pg_type_typname_nsp_index"
db_1         | 2020-06-24 15:36:34.788 UTC [110] DETAIL:  Key (typname, typnamespace)=(_SCHEMA, 2200) already exists.
db_1         | 2020-06-24 15:36:34.788 UTC [110] STATEMENT:  CREATE TABLE IF NOT EXISTS "_SCHEMA" ( "className" varChar(120), "schema" jsonb, "isParseClass" bool, PRIMARY KEY ("className") )
```

**Comment by cbaker6** [comment]
@dplewis this one is ready for review

**Comment by cbaker6** [comment]
@dplewis and @mtrezza can you review this? I believe it’s a pretty big deal for parse-server users who use Postgres and attempt to create index’s through parse-server because if a field with the same name exists in multiple tables, an index can only be created for 1 of those tables because of a bug in the code.

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 1]
```diff
@@ -2063,13 +2066,11 @@ export class PostgresStorageAdapter implements StorageAdapter {
     schema: SchemaType,
     fieldNames: string[]
   ) {
-    // Use the same name for every ensureUniqueness attempt, because postgres
```
Comment (cbaker6) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js:2066]: Here’s the main part of the bug. The assumption that was made in the comments is incorrect

**Review by davimacedo — APPROVED** [review]
LGTM!

**Pull request merged**